### PR TITLE
New vlan_mapping module

### PR DESCRIPTION
**Pull request by mingjunzhang2019**
SUMMARY
SONIC vlan mapping YANG model and design changed. This is to add sonic_vlan_mapping module design to cope with SONIC change. The resource model changed too: https://github.com/ansible-network/resource_module_models/pull/237

ISSUE TYPE
- Feature Pull Request

COMPONENT NAME
sonic_vlan_mapping

OUTPUT
- regression test result: 
[regression-2024-01-23-09-59-19.html.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14028185/regression-2024-01-23-09-59-19.html.pdf)

- regression test log: 
[Uploading regression_test.log…]()

Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

How Has This Been Tested?
Tested with real SONIC systems: regression test and unit test.

### CODE REVIEW

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 1]
```diff
@@ -143,26 +150,26 @@ def _state_replaced(self, want, have, diff):
...
+        replaced_config = self.get_replaced_config(want, have)
+
+        add_commands = []
+        if replaced_config:
+            del_requests = self.get_delete_vlan_mapping_requests(replaced_config, have,
+                                                                 is_delete_all=False,
+                                                                 state='replaced')
+            requests.extend(del_requests)
+            commands.extend(update_states(replaced_config, "deleted"))
+            add_commands = want
```
Comment (xhan-dell) [inline plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:164]: setting add_commands to want can cause extra requests if there are other mappings passed into want where want and have match.

Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:164]: Good finding! Thanks. As we redefine new "replaced" design, we will revisit "replaced" functions for all modules. So the change will be done later together with check and diff modes support.

File: plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py [hunk 2]
```diff
@@ -102,52 +102,83 @@ def get_vlan_mappings(self):
...
+                                 .get("config", {})
+                                 .get("vlan-stack-action", "SWAP"))
+
+                    if vs_action == "SWAP":
+                        vlan_trans_dict = dict()
+                        m_tag = (vlan_mapping
+                                 .get("config", {})
+                                 .get("multi-tag", False))
+                        if m_tag:
+                            vlan_trans_dict["multi_tag"] = True
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py:124]: What if m_tag is False?

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
This change set and the corresponding test results look good.

I am posting only several minor comments.

File: changelogs/fragments/296-new-vlan-mapping-resource-module.yaml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+major_changes:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply changelogs/fragments/296-new-vlan-mapping-resource-module.yaml:1]: Changed. Thanks

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 4]
```diff
@@ -322,56 +275,20 @@ def get_delete_vlan_mapping_requests(self, commands, have, is_delete_all):
...
-                            # Delete entire mapping
-                            else:
-                                path = url.format(interface_name, service_vlan)
-                                request = {"path": path, "method": method}
-                                requests.append(request)
+                if mapping_list and have_mapping_list:
+                    requests.extend(self.get_delete_vlan_mapping_mapping_requests(interface_name,
+                                                                                  mapping_list,
+                                                                                  have_mapping_list,
+                                                                                  state))
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:291]: Will revisit this later!

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 5]
```diff
@@ -381,7 +298,239 @@ def get_delete_vlan_mapping_requests(self, commands, have, is_delete_all):
...
+        requests = []
+
+        vlan_ids = dot1q_tun.get('vlan_ids', None)
+        priority = dot1q_tun.get('priority', None)
+        have_vlan_ids = have_dot1q_tun.get('vlan_ids', None)
+        have_priority = have_dot1q_tun.get('priority', None)
+
+        if vlan_ids or priority:
+            # Delete priority
+            if priority and have_priority:
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:366]: Revisit this later!

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 6]
```diff
@@ -398,70 +547,143 @@ def get_create_vlan_mapping_requests(self, commands, have):
...
-                    requests.append(self.get_create_vlan_mapping_request(interface_name, mapping))
+                    request = self.get_create_vlan_mapping_mapping_requests(interface_name,
+                                                                            mapping)
+                    if request:
+                        requests.append(request)
         return requests
 
-    def get_create_vlan_mapping_request(self, interface_name, mapping):
-        url = "data/openconfig-interfaces:interfaces/interface={}/openconfig-interfaces-ext:mapped-vlans"
+    def get_create_vlan_mapping_mapping_requests(self, interface_name, mapping):
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:556]: Revisit this later!

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 7]
```diff
@@ -398,70 +547,143 @@ def get_create_vlan_mapping_requests(self, commands, have):
...
+                        'mapped-vlan': [
+                            {
+                                'vlan-id': service_vlan,
+                                'config': {'vlan-id': service_vlan},
+                                'ingress-mapping': ing_data,
+                                'egress-mapping': egr_data
+                            }
+                        ]
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:606]: This is for readable code. But we revisit it later!

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 8]
```diff
@@ -504,14 +726,200 @@ def convert_vlan_ids_range(self, config):
...
+                mapping_index = 0
                 for mapping in mapping_list:
-                    vlan_ids = mapping.get('vlan_ids', None)
+                    if mapping.get('dot1q_tunnel', None):
+                        vlan_ids = mapping['dot1q_tunnel'].get('vlan_ids', None)
 
-                    if vlan_ids:
-                        config[interface_index]['mapping'][mapping_index]['vlan_ids'] = self.vlanIdsRangeStr(vlan_ids)
+                        if vlan_ids:
+                            dot1q_tun = config[interface_index]['mapping'][mapping_index]['dot1q_tunnel']
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:736]: Revisit this later!

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 9]
```diff
@@ -504,14 +726,200 @@ def convert_vlan_ids_range(self, config):
...
+
+    def get_replaced_vlan_mapping_mapping_translation(self, vlan_trans, have_vlan_trans):
+        rpld_trans = {}
+
+        multi_tag = vlan_trans.get('multi_tag', None)
+        ms_tags = vlan_trans.get('match_single_tags', None)
+        md_tags = vlan_trans.get('match_double_tags', None)
+
+        have_multi_tag = have_vlan_trans.get('multi_tag', False)
+
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:825]: Done. Thanks!

File: plugins/modules/sonic_vlan_mapping.py [hunk 10]
```diff
@@ -70,30 +71,73 @@
...
             description:
-              - Set priority level of the vlan mapping.
-              - Priority range is 0-7.
-            type: int
+              - Specify a vlan translation.
+            type: dict
+            suboptions:
+              multi_tag:
+                description:
+                  - Indicate if it is a multiple tags.
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_vlan_mapping.py:98]: Changed, Thanks!

File: plugins/modules/sonic_vlan_mapping.py [hunk 11]
```diff
@@ -70,30 +71,73 @@
...
+              match_single_tags:
+                description:
+                  - Configure single tagged vlan translation.
+                type: list
+                elements: dict
+                suboptions:
+                  outer_vlan:
+                    description:
+                      - Configure outer customer VLAN ID.
+                      - VLAN IDs range is 1-4094.
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_vlan_mapping.py:109]: Changed, Thanks!

File: plugins/modules/sonic_vlan_mapping.py [hunk 12]
```diff
@@ -70,30 +71,73 @@
...
+              match_double_tags:
+                description:
+                  - Configure double tagged vlan translation.
+                type: list
+                elements: dict
+                suboptions:
+                  inner_vlan:
+                    description:
+                      - Configure inner customer VLAN ID.
+                      - VLAN IDs range is 1-4094.
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_vlan_mapping.py:126]: Changed. Thanks!

File: plugins/modules/sonic_vlan_mapping.py [hunk 13]
```diff
@@ -70,30 +71,73 @@
...
+                      - Only available for double tagged translations.
+                    required: true
+                    type: int
+                  outer_vlan:
+                    description:
+                      - Configure outer customer VLAN ID.
+                      - VLAN IDs range is 1-4094.
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_vlan_mapping.py:133]: Changed. Thanks!

File: plugins/modules/sonic_vlan_mapping.py [hunk 14]
```diff
@@ -60,7 +60,8 @@
         type: str
       mapping:
         description:
-          - Defining a single vlan mapping.
+          - Define vlan mappings.
+          - dot1q_tunnel and vlan_translation are mutually exclusive.
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_vlan_mapping.py:64]: Okay. I was worried about people running the resource module builder in the future for new changes and losing the argspec 'mutually_exclusive' statement. (I thought that could be derived from the 'module' file. It isn't worth spending more time on this now, though, so I will mark this issue 'Resolved'.

File: plugins/modules/sonic_vlan_mapping.py [hunk 15]
```diff
@@ -110,393 +154,231 @@
...
-              vlan_ids:
-                - 404
-                - 401
-                - 412
-                - 430-431
-              priority: 3
-      state: deleted
-
-# After State:
+#sonic# show interface vlan-mappings
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_vlan_mapping.py:163]: It looks like the wrong line got deleted.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:366]: Okay. We can revisit this later.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:291]: Okay.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:556]: Okay.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:606]: Okay.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:736]: Okay.

File: plugins/modules/sonic_vlan_mapping.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_vlan_mapping.py]: I see that this has been fixed.

Resolved.

File: plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py (hunk 2 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py:124]: Okay.

**Review by kerry-meyer — APPROVED** [review]
All of the comments I posted have either been addressed with changes, resolved with additional explanation from the submitter, or deferred with my agreement. There are no functionality issues remaining. (The deferred comments are only for code readability.)

The current change set looks fine for merging, and the posted test results show that everything is working fine for the current version of this module.

Approved.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 1 above)
Comment (xhan-dell) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:164]: Okay with leaving this for a revisit!

**Review by xhan-dell — APPROVED** [review]

**Pull request merged**